gnmi_subscribe/gnmi_subscribe.go

```diff
@@ -77,97 +79,160 @@ func main() {
 	ctx, cancel := context.WithTimeout(context.Background(), *connectionTimeout)
 	defer cancel()
 
-	subscribeClient, err = client.Subscribe(ctx)
+	subscribeClient, err := client.Subscribe(ctx)
 	if err != nil {
 		log.Fatalf("Error creating GNMI_SubscribeClient: %v", err)
 	}
 
-	encoding, ok := pb.Encoding_value[*encodingFormat]
-	if !ok {
-		var encodingList []string
-		for _, name := range pb.Encoding_name {
-			encodingList = append(encodingList, name)
-		}
-		log.Exitf("Supported encodings: %s", strings.Join(encodingList, ", "))
+	encoding, err := parseEncoding(*encodingFormat)
+	if err != nil {
+		log.Exitf("Error parsing encoding: %v", err)
 	}
 
-	var subscriptionListMode gnmi.SubscriptionList_Mode
-	switch {
-	case *subscriptionPoll && *subscriptionOnce:
+	subscriptionListMode, err := subscriptionMode(*subscriptionPoll, *subscriptionOnce)
+	if err != nil {
 		flag.Usage()
-		log.Exitf("Only one of -once and -poll can be set")
-	case *subscriptionOnce:
-		subscriptionListMode = pb.SubscriptionList_ONCE
-	case *subscriptionPoll:
-		subscriptionListMode = pb.SubscriptionList_POLL
-	default:
-		subscriptionListMode = pb.SubscriptionList_STREAM
+		log.Exit(err)
 	}
 
-	var pbPathList []*pb.Path
-	for _, xPath := range xPathFlags {
-		pbPath, err := xpath.ToGNMIPath(xPath)
-		if err != nil {
-			log.Exitf("error in parsing xpath %q to gnmi path", xPath)
-		}
-		pbPathList = append(pbPathList, pbPath)
-	}
-	for _, textPbPath := range pbPathFlags {
-		var pbPath pb.Path
-		if err := proto.UnmarshalText(textPbPath, &pbPath); err != nil {
-			log.Exitf("error in unmarshaling %q to gnmi Path", textPbPath)
-		}
-		pbPathList = append(pbPathList, &pbPath)
+	pbPathList, err := parsePaths(xPathFlags, pbPathFlags)
+	if err != nil {
+		log.Exitf("Error parsing paths: %v", err)
 	}
-	subscriptions, err := assembleSubscriptions(pbPathList)
+
+	subscriptions, err := assembleSubscriptions(*streamOnChange, *sampleInterval, pbPathList)
 	if err != nil {
 		log.Exitf("Error assembling subscriptions: %v", err)
 	}
 
 	request := &pb.SubscribeRequest{
 		Request: &pb.SubscribeRequest_Subscribe{
 			Subscribe: &pb.SubscriptionList{
-				Encoding:     pb.Encoding(encoding),
+				Encoding:     encoding,
 				Mode:         subscriptionListMode,
 				Subscription: subscriptions,
+				UpdatesOnly:  *updatesOnly,
 			},
 		},
 	}
 	utils.PrintProto(request)
 	if err := subscribeClient.Send(request); err != nil {
 		log.Exitf("Failed to send request: %v", err)
 	}
+
 	switch subscriptionListMode {
 	case pb.SubscriptionList_STREAM:
-		stream()
+		if err := stream(subscribeClient); err != nil {
```

Comment:
I guess it makes sense to exit when using once. Would it make more sense to retry a few times with stream and poll given they could be very long running operations?

Reply:
if stream errors out then the gRPC session is broken. The retry idea is interesting but should apply at the Dial out phase.